File: packages/android-compose/src/main/kotlin/dev/chartcn/mobile/ChartSpec.kt
package dev.chartcn.mobile

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
data class ChartSpec(
  val specVersion: String,
  val id: String,
  val metadata: Metadata,
  val data: DataConfig,
  val visual: VisualConfig,
  val formatting: Formatting? = null,
  val interactions: Interactions? = null,
  val theming: Theming? = null,
  val accessibility: Accessibility,
  val platformOverrides: PlatformOverrides? = null
)

@Serializable
data class Metadata(
  val name: String,
  val description: String? = null,
  val tags: List<String> = emptyList(),
  val status: String,
  val owners: List<String>,
  val updatedAt: String? = null
)

@Serializable
data class DataConfig(
  val source: Source,
  val dimensions: List<Dimension>,
  val measures: List<Measure>,
  val filters: List<Filter> = emptyList(),
  val transforms: List<Transform> = emptyList()
)

@Serializable
data class Source(
  val adapter: Adapter,
  val entity: String? = null,
  val predicate: String? = null,
  val sort: List<OrderBy> = emptyList(),
  val limit: Int? = null,

  val table: String? = null,
  val where: String? = null,
  val args: JsonElement? = null,
  val orderBy: List<OrderBy> = emptyList(),

  val queryName: String? = null,

  val endpoint: String? = null,
  val method: String? = null,
  val headers: Map<String, String> = emptyMap(),
  val query: Map<String, JsonElement> = emptyMap(),
  val body: Map<String, JsonElement> = emptyMap(),
  val dataPath: String? = null,

  val rows: List<Map<String, JsonElement>> = emptyList()
)

@Serializable
enum class Adapter {
  @SerialName("swiftdata") SWIFTDATA,
  @SerialName("room") ROOM,
  @SerialName("sqldelight") SQLDELIGHT,
  @SerialName("api") API,
  @SerialName("static") STATIC
}

@Serializable
data class Dimension(
  val key: String,
  val type: String,
  val label: String
)

@Serializable
data class Measure(
  val key: String,
  val type: String,
  val label: String,
  val unit: String? = null,
  val currency: String? = null
)

@Serializable
data class Filter(
  val field: String,
  val op: String,
  val value: JsonElement
)

@Serializable
data class Transform(
  val type: String,
  val by: String? = null,
  val direction: String? = null,
  val input: String? = null,
  val window: Int? = null,
  val `as`: String? = null,
  val aggregations: List<Aggregation> = emptyList()
)

@Serializable
data class Aggregation(
  val field: String,
  val op: String,
  val `as`: String
)

@Serializable
data class OrderBy(
  val field: String,
  val direction: String
)

@Serializable
data class VisualConfig(
  val chartType: ChartType,
  val xField: String? = null,
  val groupField: String? = null,
  val stacked: Boolean = false,
  val series: List<Series>,
  val legend: Legend? = null,
  val axes: Axes? = null,
  val tooltip: Tooltip? = null,
  val emptyState: EmptyState? = null
)

@Serializable
enum class ChartType {
  @SerialName("line") LINE,
  @SerialName("bar") BAR,
  @SerialName("area") AREA,
  @SerialName("pie") PIE,
  @SerialName("donut") DONUT,
  @SerialName("scatter") SCATTER,
  @SerialName("combo") COMBO,
  @SerialName("kpi") KPI
}

@Serializable
data class Series(
  val field: String,
  val label: String,
  val renderer: String? = null,
  val style: SeriesStyle? = null
)

@Serializable
data class SeriesStyle(
  val color: String? = null,
  val lineWidth: Double? = null,
  val dash: List<Double> = emptyList(),
  val opacity: Double? = null
)

@Serializable
data class Legend(
  val visible: Boolean = true,
  val position: String? = null
)

@Serializable
data class Axes(
  val x: Axis? = null,
  val y: Axis? = null
)

@Serializable
data class Axis(
  val label: String? = null,
  val min: Double? = null,
  val max: Double? = null,
  val tickCount: Int? = null
)

@Serializable
data class Tooltip(
  val enabled: Boolean = true,
  val mode: String? = null
)

@Serializable
data class EmptyState(
  val title: String? = null,
  val description: String? = null
)

@Serializable
data class Formatting(
  val number: NumberFormatting? = null,
  val currency: CurrencyFormatting? = null,
  val date: DateFormatting? = null
)

@Serializable
data class NumberFormatting(
  val notation: String? = null,
  val maximumFractionDigits: Int? = null
)

@Serializable
data class CurrencyFormatting(
  val code: String? = null,
  val display: String? = null
)

@Serializable
data class DateFormatting(
  val granularity: String? = null
)

@Serializable
data class Interactions(
  val selection: String? = null,
  val drilldown: Drilldown? = null,
  val gestures: List<String> = emptyList()
)

@Serializable
data class Drilldown(
  val enabled: Boolean = false,
  val targetRoute: String? = null,
  val paramField: String? = null
)

@Serializable
data class Theming(
  val palette: String? = null,
  val tokens: Map<String, String> = emptyMap()
)

@Serializable
data class Accessibility(
  val chartTitle: String,
  val summaryTemplate: String,
  val announceOnLoad: Boolean = true
)

@Serializable
data class PlatformOverrides(
  val ios: IOSOverrides? = null,
  val android: AndroidOverrides? = null
)

@Serializable
data class IOSOverrides(
  val interpolation: String? = null,
  val symbolSize: Double? = null
)

@Serializable
data class AndroidOverrides(
  val curveStyle: String? = null,
  val pointRadius: Double? = null
)

typealias ChartRow = Map<String, JsonElement>

File: packages/ios-swiftui/Sources/ChartCNMobile/ChartSpec.swift
import Foundation

public enum ChartType: String, Codable, Sendable {
    case line
    case bar
    case area
    case pie
    case donut
    case scatter
    case combo
    case kpi
}

public enum ChartAdapterType: String, Codable, Sendable {
    case swiftdata
    case room
    case sqldelight
    case api
    case `static`
}

public enum ChartValue: Codable, Sendable, Hashable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case array([ChartValue])
    case object([String: ChartValue])
    case null

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()

        if container.decodeNil() {
            self = .null
            return
        }

        if let bool = try? container.decode(Bool.self) {
            self = .bool(bool)
            return
        }

        if let number = try? container.decode(Double.self) {
            self = .number(number)
            return
        }

        if let string = try? container.decode(String.self) {
            self = .string(string)
            return
        }

        if let array = try? container.decode([ChartValue].self) {
            self = .array(array)
            return
        }

        if let object = try? container.decode([String: ChartValue].self) {
            self = .object(object)
            return
        }

        throw DecodingError.typeMismatch(
            ChartValue.self,
            DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Unsupported ChartValue type")
        )
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()

        switch self {
        case .string(let value):
            try container.encode(value)
        case .number(let value):
            try container.encode(value)
        case .bool(let value):
            try container.encode(value)
        case .array(let value):
            try container.encode(value)
        case .object(let value):
            try container.encode(value)
        case .null:
            try container.encodeNil()
        }
    }

    public var stringValue: String? {
        if case .string(let value) = self { return value }
        return nil
    }

    public var doubleValue: Double? {
        if case .number(let value) = self { return value }
        return nil
    }

    public var arrayValue: [ChartValue]? {
        if case .array(let value) = self { return value }
        return nil
    }

    public var objectValue: [String: ChartValue]? {
        if case .object(let value) = self { return value }
        return nil
    }
}

public typealias ChartRow = [String: ChartValue]

public struct ChartSpec: Codable, Sendable {
    public let specVersion: String
    public let id: String
    public let metadata: Metadata
    public let data: DataConfig
    public let visual: VisualConfig
    public let formatting: Formatting?
    public let interactions: Interactions?
    public let theming: Theming?
    public let accessibility: Accessibility
    public let platformOverrides: PlatformOverrides?

    public struct Metadata: Codable, Sendable {
        public let name: String
        public let description: String?
        public let tags: [String]?
        public let status: String
        public let owners: [String]
        public let updatedAt: String?
    }

    public struct DataConfig: Codable, Sendable {
        public let source: Source
        public let dimensions: [Dimension]
        public let measures: [Measure]
        public let filters: [Filter]?
        public let transforms: [Transform]?
    }

    public struct Source: Codable, Sendable {
        public let adapter: ChartAdapterType
        public let entity: String?
        public let predicate: String?
        public let sort: [OrderBy]?
        public let limit: Int?

        public let table: String?
        public let `where`: String?
        public let args: ChartValue?
        public let orderBy: [OrderBy]?

        public let queryName: String?

        public let endpoint: String?
        public let method: String?
        public let headers: [String: String]?
        public let query: [String: ChartValue]?
        public let body: [String: ChartValue]?
        public let dataPath: String?

        public let rows: [ChartRow]?
    }

    public struct OrderBy: Codable, Sendable {
        public let field: String
        public let direction: String
    }

    public struct Dimension: Codable, Sendable {
        public let key: String
        public let type: String
        public let label: String
    }

    public struct Measure: Codable, Sendable {
        public let key: String
        public let type: String
        public let label: String
        public let unit: String?
        public let currency: String?
    }

    public struct Filter: Codable, Sendable {
        public let field: String
        public let op: String
        public let value: ChartValue
    }

    public struct Transform: Codable, Sendable {
        public let type: String
        public let by: String?
        public let direction: String?
        public let input: String?
        public let window: Int?
        public let `as`: String?
        public let aggregations: [Aggregation]?

        public struct Aggregation: Codable, Sendable {
            public let field: String
            public let op: String
            public let `as`: String
        }
    }

    public struct VisualConfig: Codable, Sendable {
        public let chartType: ChartType
        public let xField: String?
        public let groupField: String?
        public let stacked: Bool?
        public let series: [Series]
        public let legend: Legend?
        public let axes: Axes?
        public let tooltip: Tooltip?
        public let emptyState: EmptyState?

        public struct Series: Codable, Sendable {
            public let field: String
            public let label: String
            public let renderer: String?
            public let style: Style?

            public struct Style: Codable, Sendable {
                public let color: String?
                public let lineWidth: Double?
                public let dash: [Double]?
                public let opacity: Double?
            }
        }

        public struct Legend: Codable, Sendable {
            public let visible: Bool?
            public let position: String?
        }

        public struct Axes: Codable, Sendable {
            public let x: Axis?
            public let y: Axis?

            public struct Axis: Codable, Sendable {
                public let label: String?
                public let min: Double?
                public let max: Double?
                public let tickCount: Int?
            }
        }

        public struct Tooltip: Codable, Sendable {
            public let enabled: Bool?
            public let mode: String?
        }

        public struct EmptyState: Codable, Sendable {
            public let title: String?
            public let description: String?
        }
    }

    public struct Formatting: Codable, Sendable {
        public let number: NumberFormatting?
        public let currency: CurrencyFormatting?
        public let date: DateFormatting?

        public struct NumberFormatting: Codable, Sendable {
            public let notation: String?
            public let maximumFractionDigits: Int?
        }

        public struct CurrencyFormatting: Codable, Sendable {
            public let code: String?
            public let display: String?
        }

        public struct DateFormatting: Codable, Sendable {
            public let granularity: String?
        }
    }

    public struct Interactions: Codable, Sendable {
        public let selection: String?
        public let drilldown: Drilldown?
        public let gestures: [String]?

        public struct Drilldown: Codable, Sendable {
            public let enabled: Bool?
            public let targetRoute: String?
            public let paramField: String?
        }
    }

    public struct Theming: Codable, Sendable {
        public let palette: String?
        public let tokens: [String: String]?
    }

    public struct Accessibility: Codable, Sendable {
        public let chartTitle: String
        public let summaryTemplate: String
        public let announceOnLoad: Bool?
    }

    public struct PlatformOverrides: Codable, Sendable {
        public let ios: IOSOverrides?
        public let android: AndroidOverrides?

        public struct IOSOverrides: Codable, Sendable {
            public let interpolation: String?
            public let symbolSize: Double?
        }

        public struct AndroidOverrides: Codable, Sendable {
            public let curveStyle: String?
            public let pointRadius: Double?
        }
    }
}
